File: drivers/go/redwire/scram.go
package redwire

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// SCRAM-SHA-256 client primitives. RFC 5802 + RFC 7677.
//
// Hand-rolled HMAC / PBKDF2 to keep the dependency tree minimal, even though
// crypto/hmac and golang.org/x/crypto/pbkdf2 exist — staying inline mirrors
// drivers/rust/src/redwire/scram.rs and lets us test against RFC vectors
// without an external crate.

// DefaultIter mirrors the engine's RFC 7677 default.
const DefaultIter = 16384

// MinIter is the lowest iteration count this client will agree to use.
const MinIter = 4096

// HMACSHA256 returns HMAC-SHA-256(key, data). Implemented inline to match the
// Rust client byte-for-byte.
func HMACSHA256(key, data []byte) [32]byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(data)
	var out [32]byte
	copy(out[:], mac.Sum(nil))
	return out
}

// SHA256 returns SHA-256(data).
func SHA256(data []byte) [32]byte {
	return sha256.Sum256(data)
}

// PBKDF2SHA256 derives a 32-byte key. Iter must be ≥ 1; callers should
// pass MinIter or higher.
func PBKDF2SHA256(password, salt []byte, iter uint32) [32]byte {
	// Block index always 1 because we want exactly one 32-byte block.
	salted := make([]byte, 0, len(salt)+4)
	salted = append(salted, salt...)
	idx := make([]byte, 4)
	binary.BigEndian.PutUint32(idx, 1)
	salted = append(salted, idx...)
	u := HMACSHA256(password, salted)
	out := u
	for i := uint32(1); i < iter; i++ {
		u = HMACSHA256(password, u[:])
		for j := 0; j < 32; j++ {
			out[j] ^= u[j]
		}
	}
	return out
}

// XOR returns a^b. Inputs must be the same length.
func XOR(a, b []byte) []byte {
	if len(a) != len(b) {
		return nil
	}
	out := make([]byte, len(a))
	for i := range a {
		out[i] = a[i] ^ b[i]
	}
	return out
}

// AuthMessage builds the canonical RFC 5802 §3 auth-message string:
//
//	client-first-bare + "," + server-first + "," + client-final-no-proof
func AuthMessage(clientFirstBare, serverFirst, clientFinalNoProof string) []byte {
	return []byte(clientFirstBare + "," + serverFirst + "," + clientFinalNoProof)
}

// ClientProof — what the client sends to prove it knows the password.
//
//	salted    = PBKDF2-HMAC-SHA256(password, salt, iter, 32)
//	clientKey = HMAC-SHA256(salted, "Client Key")
//	storedKey = SHA256(clientKey)
//	signature = HMAC-SHA256(storedKey, authMessage)
//	proof     = clientKey XOR signature
func ClientProof(password, salt []byte, iter uint32, authMessage []byte) []byte {
	salted := PBKDF2SHA256(password, salt, iter)
	clientKey := HMACSHA256(salted[:], []byte("Client Key"))
	storedKey := SHA256(clientKey[:])
	sig := HMACSHA256(storedKey[:], authMessage)
	return XOR(clientKey[:], sig[:])
}

// VerifyServerSignature checks the `v=` field in AuthOk against the recomputed
// HMAC-SHA-256(serverKey, authMessage). Constant-time comparison.
func VerifyServerSignature(password, salt []byte, iter uint32, authMessage, presented []byte) bool {
	if len(presented) != 32 {
		return false
	}
	salted := PBKDF2SHA256(password, salt, iter)
	serverKey := HMACSHA256(salted[:], []byte("Server Key"))
	expected := HMACSHA256(serverKey[:], authMessage)
	return subtle.ConstantTimeCompare(expected[:], presented) == 1
}

// NewClientNonce returns a fresh 24-byte client nonce, base64-std encoded
// without padding stripped (matches what the Rust driver and the engine emit).
func NewClientNonce() (string, error) {
	var raw [18]byte
	if _, err := rand.Read(raw[:]); err != nil {
		return "", fmt.Errorf("redwire scram: random: %w", err)
	}
	return base64.StdEncoding.EncodeToString(raw[:]), nil
}

// ScramServerFirst is the parsed `r=...,s=...,i=...` payload the server sends
// in its AuthRequest after we send client-first.
type ScramServerFirst struct {
	CombinedNonce string
	Salt          []byte
	Iter          uint32
	Raw           string // original textual form, used to build authMessage
}

// ParseServerFirst parses the server-first message bytes.
func ParseServerFirst(raw []byte) (*ScramServerFirst, error) {
	s := string(raw)
	out := &ScramServerFirst{Raw: s}
	for _, part := range strings.Split(s, ",") {
		switch {
		case strings.HasPrefix(part, "r="):
			out.CombinedNonce = part[2:]
		case strings.HasPrefix(part, "s="):
			salt, err := base64.StdEncoding.DecodeString(part[2:])
			if err != nil {
				return nil, fmt.Errorf("redwire scram: bad salt: %w", err)
			}
			out.Salt = salt
		case strings.HasPrefix(part, "i="):
			n, err := strconv.ParseUint(part[2:], 10, 32)
			if err != nil {
				return nil, fmt.Errorf("redwire scram: bad iter: %w", err)
			}
			out.Iter = uint32(n)
		}
	}
	if out.CombinedNonce == "" {
		return nil, errors.New("redwire scram: server-first missing r=")
	}
	if len(out.Salt) == 0 {
		return nil, errors.New("redwire scram: server-first missing s=")
	}
	if out.Iter == 0 {
		return nil, errors.New("redwire scram: server-first missing i=")
	}
	return out, nil
}

// EncodeBase64Std returns base64-std encoding of input (with `=` padding).
func EncodeBase64Std(input []byte) string {
	return base64.StdEncoding.EncodeToString(input)
}

// DecodeBase64Std accepts both padded and unpadded forms.
func DecodeBase64Std(input string) ([]byte, error) {
	if strings.HasSuffix(input, "=") {
		return base64.StdEncoding.DecodeString(input)
	}
	return base64.RawStdEncoding.DecodeString(input)
}

// ScramSession packages the client-side state needed to drive a SCRAM exchange
// across the three RTTs in the RedWire handshake.
type ScramSession struct {
	Username        string
	Password        []byte
	ClientNonce     string
	ClientFirstBare string
}

// NewScramSession seeds a session with a fresh nonce.
func NewScramSession(username, password string) (*ScramSession, error) {
	nonce, err := NewClientNonce()
	if err != nil {
		return nil, err
	}
	return &ScramSession{
		Username:        username,
		Password:        []byte(password),
		ClientNonce:     nonce,
		ClientFirstBare: "n=" + username + ",r=" + nonce,
	}, nil
}

// ClientFirstMessage returns the GS2 + bare form to send in AuthResponse.
func (s *ScramSession) ClientFirstMessage() string {
	return "n,," + s.ClientFirstBare
}

// BuildClientFinal computes the client-final message + auth-message bytes
// based on the parsed server-first.
func (s *ScramSession) BuildClientFinal(sf *ScramServerFirst) (final string, authMessage []byte, err error) {
	if !strings.HasPrefix(sf.CombinedNonce, s.ClientNonce) {
		return "", nil, errors.New("redwire scram: server combined nonce does not include our client nonce")
	}
	if sf.Iter < MinIter {
		return "", nil, fmt.Errorf("redwire scram: iter %d below minimum %d", sf.Iter, MinIter)
	}
	clientFinalNoProof := "c=biws,r=" + sf.CombinedNonce
	authMessage = AuthMessage(s.ClientFirstBare, sf.Raw, clientFinalNoProof)
	proof := ClientProof(s.Password, sf.Salt, sf.Iter, authMessage)
	final = clientFinalNoProof + ",p=" + EncodeBase64Std(proof)
	return final, authMessage, nil
}

File: drivers/dart/lib/src/redwire/scram.dart
/// SCRAM-SHA-256 client primitives (RFC 5802 / RFC 7677).
///
/// Pure functions, no I/O. Mirrors `drivers/rust/src/redwire/scram.rs`
/// and the Python driver so all three are bit-for-bit interoperable
/// with the engine's `src/auth/scram.rs`.
library;

import 'dart:convert';
import 'dart:math' show Random;
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

const int defaultIter = 16384;
const int minIter = 4096;

Uint8List hmacSha256(List<int> key, List<int> data) {
  final mac = Hmac(sha256, key).convert(data);
  return Uint8List.fromList(mac.bytes);
}

Uint8List sha256Of(List<int> data) {
  return Uint8List.fromList(sha256.convert(data).bytes);
}

/// PBKDF2-HMAC-SHA256 with a fixed 32-byte derived key.
///
/// Hand-rolled because we don't want to drag in `package:cryptography`
/// just for this. Loop count `iterations` matches RFC 8018 / 6070.
Uint8List pbkdf2Sha256(
  List<int> password,
  List<int> salt,
  int iterations, {
  int dkLen = 32,
}) {
  if (iterations < 1) {
    throw ArgumentError.value(iterations, 'iterations', 'must be >= 1');
  }
  // For dkLen <= 32 we only need block index 1.
  final blockCount = (dkLen + 31) ~/ 32;
  final out = BytesBuilder(copy: false);
  for (var i = 1; i <= blockCount; i++) {
    final saltedBlock = Uint8List(salt.length + 4);
    saltedBlock.setRange(0, salt.length, salt);
    saltedBlock[salt.length] = (i >> 24) & 0xFF;
    saltedBlock[salt.length + 1] = (i >> 16) & 0xFF;
    saltedBlock[salt.length + 2] = (i >> 8) & 0xFF;
    saltedBlock[salt.length + 3] = i & 0xFF;
    var u = hmacSha256(password, saltedBlock);
    final block = Uint8List.fromList(u);
    for (var iter = 1; iter < iterations; iter++) {
      u = hmacSha256(password, u);
      for (var j = 0; j < 32; j++) {
        block[j] ^= u[j];
      }
    }
    out.add(block);
  }
  final all = out.toBytes();
  return all.length == dkLen ? all : Uint8List.sublistView(all, 0, dkLen);
}

Uint8List xorBytes(List<int> a, List<int> b) {
  if (a.length != b.length) {
    throw ArgumentError('xor inputs must be equal length');
  }
  final out = Uint8List(a.length);
  for (var i = 0; i < a.length; i++) {
    out[i] = a[i] ^ b[i];
  }
  return out;
}

/// Standard-alphabet base64 with padding — matches the server's
/// `base64_std`.
String b64encode(List<int> data) => base64.encode(data);

/// Standard-alphabet base64 decode. Tolerant of missing padding.
Uint8List b64decode(String text) {
  final pad = (-text.length) % 4;
  return Uint8List.fromList(base64.decode(text + ('=' * pad)));
}

/// 24-character base64 nonce produced from 18 random bytes.
///
/// Uses [Random.secure] when available; falls back to [Random] only if
/// the platform refuses (very unusual).
String makeClientNonce({int numBytes = 18, Random? random}) {
  final rng = random ?? Random.secure();
  final raw = Uint8List(numBytes);
  for (var i = 0; i < numBytes; i++) {
    raw[i] = rng.nextInt(256);
  }
  return b64encode(raw);
}

/// Build the client-first message and its bare form.
///
/// Returns `(client_first_message, client_first_bare)`. The bare form
/// is needed for the auth message later.
({String full, String bare}) buildClientFirst(String username, String clientNonce) {
  final bare = 'n=$username,r=$clientNonce';
  return (full: 'n,,$bare', bare: bare);
}

/// Parsed server-first message.
class ServerFirst {
  ServerFirst({
    required this.combinedNonce,
    required this.salt,
    required this.iterations,
  });

  final String combinedNonce;
  final Uint8List salt;
  final int iterations;
}

ServerFirst parseServerFirst(String serverFirst, String clientNonce) {
  String? combined;
  String? saltB64;
  int? iters;
  for (final part in serverFirst.split(',')) {
    if (part.startsWith('r=')) {
      combined = part.substring(2);
    } else if (part.startsWith('s=')) {
      saltB64 = part.substring(2);
    } else if (part.startsWith('i=')) {
      final n = int.tryParse(part.substring(2));
      if (n == null) {
        throw FormatException("invalid iter in server-first: '$part'");
      }
      iters = n;
    }
  }
  if (combined == null || saltB64 == null || iters == null) {
    throw FormatException("server-first missing fields: '$serverFirst'");
  }
  if (!combined.startsWith(clientNonce)) {
    throw FormatException(
      'server nonce does not start with client nonce (replay protection)',
    );
  }
  if (iters < minIter) {
    throw FormatException(
      'server-supplied iter ($iters) is below MIN_ITER=$minIter',
    );
  }
  return ServerFirst(
    combinedNonce: combined,
    salt: b64decode(saltB64),
    iterations: iters,
  );
}

/// Concatenate `client_first_bare,server_first,client_final_no_proof`.
Uint8List authMessage(
  String clientFirstBare,
  String serverFirst,
  String clientFinalNoProof,
) {
  return Uint8List.fromList(
    utf8.encode('$clientFirstBare,$serverFirst,$clientFinalNoProof'),
  );
}

/// Compute `ClientKey XOR HMAC(StoredKey, AuthMessage)` (32 bytes).
Uint8List clientProof(
  List<int> password,
  List<int> salt,
  int iterations,
  List<int> am,
) {
  final salted = pbkdf2Sha256(password, salt, iterations);
  final clientKey = hmacSha256(salted, utf8.encode('Client Key'));
  final storedKey = sha256Of(clientKey);
  final sig = hmacSha256(storedKey, am);
  return xorBytes(clientKey, sig);
}

/// Build the client-final message and the no-proof prefix.
({String full, String noProof}) buildClientFinal(
  String combinedNonce,
  Uint8List proof,
) {
  final noProof = 'c=biws,r=$combinedNonce';
  return (full: '$noProof,p=${b64encode(proof)}', noProof: noProof);
}

Uint8List serverSignature(
  List<int> password,
  List<int> salt,
  int iterations,
  List<int> am,
) {
  final salted = pbkdf2Sha256(password, salt, iterations);
  final serverKey = hmacSha256(salted, utf8.encode('Server Key'));
  return hmacSha256(serverKey, am);
}

bool constantTimeEq(List<int> a, List<int> b) {
  if (a.length != b.length) return false;
  var diff = 0;
  for (var i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff == 0;
}

bool verifyServerSignature(
  List<int> password,
  List<int> salt,
  int iterations,
  List<int> am,
  List<int> presented,
) {
  if (presented.length != 32) return false;
  final expected = serverSignature(password, salt, iterations, am);
  return constantTimeEq(expected, presented);
}
